File: projeto1/src/matrix.cpp
#include <stdio.h>
#include <iostream>
#include <fstream>
#include <iomanip>
#include <time.h>
#include <cstdlib>
#include <papi.h>

using namespace std;

#define SYSTEMTIME clock_t

 
void OnMult(int m_ar, int m_br, FILE* data) 
{
    fprintf(data, "Multiplication, %d, -, ", m_ar);
	SYSTEMTIME Time1, Time2;
	double temp;
	int i, j, k;

	double *pha, *phb, *phc;
	
    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
	phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
	phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

	for(i=0; i<m_ar; i++)
		for(j=0; j<m_ar; j++)
			pha[i*m_ar + j] = (double)1.0;


	for(i=0; i<m_br; i++)
		for(j=0; j<m_br; j++)
			phb[i*m_br + j] = (double)(i+1);


    Time1 = clock();

	for(i=0; i<m_ar; i++) {
		for( j=0; j<m_br; j++) {
			temp = 0;
			for( k=0; k<m_ar; k++) {	
				temp += pha[i*m_ar+k] * phb[k*m_br+j];
			}
			phc[i*m_ar+j]=temp;
		}
	}

    Time2 = clock();

	fprintf(data, "%3.3f, ", (double)(Time2 - Time1) / CLOCKS_PER_SEC);

	// display 10 elements of the result matrix tto verify correctness
	cout << "Result matrix: " << endl;
	for(i=0; i<1; i++)
	{	for(j=0; j<min(10,m_br); j++)
			cout << phc[j] << " ";
	}
	cout << endl;

    free(pha);
    free(phb);
    free(phc);
}

// add code here for line x line matriz multiplication
void OnMultLine(int m_ar, int m_br, FILE* data)
{
    SYSTEMTIME Time1, Time2;
    fprintf(data, "Line Multiplication, %d, -, ", m_ar);

    int i, j, k;

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

    for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = (double)(i + 1);

    Time1 = clock();

    for (i = 0; i < m_ar; i++){
        for (k = 0; k < m_ar; k++){
            for (j = 0; j < m_br; j++){
                phc[i*m_ar+j] += pha[i*m_ar+j] * phb[k*m_br+j];
            }
        }
    }

    Time2 = clock();

	fprintf(data, "%3.3f, ", (double)(Time2 - Time1) / CLOCKS_PER_SEC);

    // display 10 elements of the result matrix tto verify correctness
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++){
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
    
}

void OnMultBlock(int m_ar, int m_br, int bkSize, FILE* data)
{
	SYSTEMTIME Time1, Time2;
	fprintf(data, "Block Multiplication, %d, %d, ", m_ar, bkSize);

    double *pha, *phb, *phc;

    pha = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phb = (double *)malloc((m_ar * m_ar) * sizeof(double));
    phc = (double *)malloc((m_ar * m_ar) * sizeof(double));

	int i, j;

	for (i = 0; i < m_ar; i++)
        for (j = 0; j < m_ar; j++)
            pha[i * m_ar + j] = (double)1.0;

    for (i = 0; i < m_br; i++)
        for (j = 0; j < m_br; j++)
            phb[i * m_br + j] = (double)(i + 1);

    Time1 = clock();

	int k;
	for (int bi = 0; bi < m_ar; bi += bkSize){
        for (int bj = 0; bj < m_br; bj += bkSize){
            for (int bk = 0; bk < m_ar; bk += bkSize){
                // Line mult
                for (i = 0; i < bkSize; i++){
                    for (k = 0; k < bkSize; k++){
                        for (j = 0; j < bkSize; j++){
                            phc[(bi + i) * m_ar + bj + j] += pha[(bi + i) * m_ar + bk + k] * phb[(bk + k) * m_ar + bj + j];
                        }
                    }
                }
            }
        }
    }

    Time2 = clock();
	fprintf(data, "%3.3f, ", (double)(Time2 - Time1) / CLOCKS_PER_SEC);

    // display 10 elements of the result matrix tto verify correctness
    cout << "Result matrix: " << endl;
    for (i = 0; i < 1; i++){
        for (j = 0; j < min(10, m_br); j++)
            cout << phc[j] << " ";
    }
    cout << endl;

    free(pha);
    free(phb);
    free(phc);
}

void handle_error (int retval)
{
  printf("PAPI error %d: %s\n", retval, PAPI_strerror(retval));
  exit(1);
}

void init_papi() {
  int retval = PAPI_library_init(PAPI_VER_CURRENT);
  if (retval != PAPI_VER_CURRENT && retval < 0) {
    printf("PAPI library version mismatch!\n");
    exit(1);
  }
  if (retval < 0) handle_error(retval);

  std::cout << "PAPI Version Number: MAJOR: " << PAPI_VERSION_MAJOR(retval)
            << " MINOR: " << PAPI_VERSION_MINOR(retval)
            << " REVISION: " << PAPI_VERSION_REVISION(retval) << "\n";
}

int main (int argc, char *argv[])
{
	FILE* data;

	data = fopen("data_cpp.csv", "w");
    fprintf(data, "type, size, block_size, time, L1_DCM, L2_DCM, TOT_CYC\n");

  	if (data == NULL) {
		perror("Cannot open file.");
	}

	int lin, col, blockSize;
	int operation;
	
	int EventSet = PAPI_NULL;
  	long long values[3];
  	int ret;
	

	ret = PAPI_library_init( PAPI_VER_CURRENT );
	if ( ret != PAPI_VER_CURRENT )
		std::cout << "FAIL" << endl;


	ret = PAPI_create_eventset(&EventSet);
	if (ret != PAPI_OK) cout << "ERROR: create eventset" << endl;


	ret = PAPI_add_event(EventSet,PAPI_L1_DCM );
	if (ret != PAPI_OK) cout << "ERROR: PAPI_L1_DCM" << endl;


	ret = PAPI_add_event(EventSet,PAPI_L2_DCM);
	if (ret != PAPI_OK) cout << "ERROR: PAPI_L2_DCM" << endl;

	ret = PAPI_add_event(EventSet,PAPI_TOT_CYC);
	if (ret != PAPI_OK) cout << "ERROR: PAPI_TOT_CYC" << endl;

	operation = 1;

	do {
		cout << endl << "1. Multiplication" << endl;
		cout << "2. Line Multiplication" << endl;
		cout << "3. Block Multiplication" << endl;
		cout << "0. Exit" << endl;
		cout << "Selection?: ";
		cin >> operation;
		if (operation == 0)
			break;
		printf("Dimensions: lins=cols ? ");
   		cin >> lin;
   		col = lin;

		// Start counting
		ret = PAPI_start(EventSet);
		if (ret != PAPI_OK) cout << "ERROR: Start PAPI" << endl;

		switch (operation){
			case 1: 
				OnMult(lin, col, data);
				break;
			case 2:
				OnMultLine(lin, col, data);  
				break;
			case 3:
				cout << "Block Size? ";
				cin >> blockSize;
				OnMultBlock(lin, col, blockSize, data);  
				break;
		}

  		ret = PAPI_stop(EventSet, values);
  		if (ret != PAPI_OK) cout << "ERROR: Stop PAPI" << endl;

		fprintf(data, "%lld, %lld, %lld\n",values[0], values[1], values[2]);

		ret = PAPI_reset( EventSet );
		if ( ret != PAPI_OK )
			std::cout << "FAIL reset" << endl; 

	}while (operation != 0);

	ret = PAPI_remove_event( EventSet, PAPI_L1_DCM );
	if ( ret != PAPI_OK )
		std::cout << "FAIL remove event" << endl; 

	ret = PAPI_remove_event( EventSet, PAPI_L2_DCM );
	if ( ret != PAPI_OK )
		std::cout << "FAIL remove event" << endl; 

	ret = PAPI_destroy_eventset( &EventSet );
	if ( ret != PAPI_OK )
		std::cout << "FAIL destroy" << endl;

	fclose(data);
}

File: projeto1/src/matrix.java
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner; 


public class matrix {

    private static final double CLOCKS_PER_SEC = 1000000000;

    static FileWriter fileWriter;
    static PrintWriter printWriter;

    static FileWriter fileWriterMD;
    static PrintWriter printWriterMD;

    public static void OnMult(int m_ar, int m_br){

        printWriter.printf("Multiplication, %d, -, ", m_ar);
        long startTime, endTime,duration;
        
        double temp;
        int i, j, k;
    
        
        double pha[] = new double[m_ar * m_ar];
        double phb[] = new double[m_ar * m_ar];
        double phc[] = new double[m_ar * m_ar];

    
        for(i=0; i<m_ar; i++)
            for(j=0; j<m_ar; j++)
                pha[i*m_ar + j] = (double)1.0;
    
    
        for(i=0; i<m_br; i++)
            for(j=0; j<m_br; j++)
                phb[i*m_br + j] = (double)(i+1);
    
        startTime = System.nanoTime();
        
        for(i=0; i<m_ar; i++) {
            for( j=0; j<m_br; j++) {
                temp = 0;
                for( k=0; k<m_ar; k++) {	
                    temp += pha[i*m_ar+k] * phb[k*m_br+j];
                }
                phc[i*m_ar+j]=temp;
            }
        }
    
        endTime = System.nanoTime();
        duration = (endTime - startTime);
        
        System.out.println("Time: " + duration/ CLOCKS_PER_SEC + " seconds");

        printWriter.printf("%3.3f\n", duration / CLOCKS_PER_SEC);
        
        // display 10 elements of the result matrix tto verify correctness
        System.out.println("Result matrix: ");
        for(i=0; i<1; i++)
        {	for(j=0; j< Math.min(10,m_br); j++)
                System.out.print(phc[j] + " ");
        }
        System.out.println();
    }
    public static void OnMultLine(int m_ar, int m_br){

        printWriter.printf("Line Multiplication, %d, -, ", m_ar);

        long startTime, endTime, duration;
        
        int i, j, k;
    
        double pha[] = new double[m_ar * m_ar];
        double phb[] = new double[m_ar * m_ar];
        double phc[] = new double[m_ar * m_ar];

        for(i=0; i<m_ar; i++)
            for(j=0; j<m_ar; j++)
                pha[i*m_ar + j] = (double)1.0;
    
    
        for(i=0; i<m_br; i++)
            for(j=0; j<m_br; j++)
                phb[i*m_br + j] = (double)(i+1);
    
        startTime = System.nanoTime();
        
        for (i = 0; i < m_ar; i++){
            for (k = 0; k < m_ar; k++){
                for (j = 0; j < m_br; j++){
                    phc[i*m_ar+j] += pha[i*m_ar+j] * phb[k*m_br+j];
                }
            }
        }
    
        endTime = System.nanoTime();
        duration = (endTime - startTime);

        System.out.println("Time: " + duration/CLOCKS_PER_SEC + " seconds");

        printWriter.printf("%3.3f, ", duration / CLOCKS_PER_SEC);
    
        // display 10 elements of the result matrix tto verify correctness
        System.out.println("Result matrix: ");
        for(i=0; i<1; i++)
        {	for(j=0; j< Math.min(10,m_br); j++)
                System.out.print(phc[j] + " ");
        }
        System.out.println();
    }
   
    public static void main(String[] args){
        Scanner input = new Scanner(System.in); 
        int operation = 0;
        int lin, col;

        try {
            fileWriter = new FileWriter("data_java.csv");
            printWriter = new PrintWriter(fileWriter);

	        printWriter.printf("type, size, block_size, time, L1_DCM, L2_DCM, TOT_CYC\n");
            
        } catch (IOException e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }

        do{
            System.out.println("1. Multiplication");
            System.out.println("2. Line Multiplication");
            System.out.println("0. Exit");
            System.out.println("Selection?: ");

            operation = input.nextInt();
            if (operation == 0)
			    break;

            System.out.println("Dimensions: lins=cols ? ");
            lin = input.nextInt();
            col=lin;

            switch (operation){
                case 1: 
                    OnMult(lin, col);
                    break;
                case 2:
                    OnMultLine(lin, col);  
                    break;
            }
        } while (operation != 0);
    
        
        try {
            fileWriter.close();
            input.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }
}